WIP: Use native go library

SantiagoTorres opened the pull request:
I thought this was already part of the main tree. This PR adds:

- Support for the native go implementation
- Fixes for unused imports

Kudos to @AdityaSaky for the initial stab at this!

adityasaky committed:
Initial implmentation of switch to native golang verify

adityasaky committed:
Remove unused import os/exec

SantiagoTorres committed:
main: update usage message, remove unused imports

SantiagoTorres committed:
main:gopkg: fix imports

SantiagoTorres committed:
main:in_toto: use native go implementation

SantiagoTorres committed:
test: add empty test harness

SantiagoTorres committed:
ENH: move to go modules

SantiagoTorres committed:
example: add example directory

SantiagoTorres committed:
README: add example mention and update usage

SantiagoTorres committed:
updepends: go.sum ++, go.mod ++

SantiagoTorres committed:
main, in_toto: delineate kubectl vs k8s api funcs

The functions useful for other k8s-related applications (e.g., an
admission controller) were moved to the in_toto pkg, whereas all of the
kubectl related functionalities were kept on main. This way, we can
simplify both codebases by reusing more code...

lukpueh commented:
This is cool stuff, @SantiagoTorres! Haven't looked at the plugin code yet, but I skimmed through the example and I really like it. Here are a few suggestions/questions:

- A single copy-pastable demo instructions-`.md` would be nice 
- What's the recommended way to start the deployment? Is it `kubectl apply -f example/deployments/test.yml`?
- After running above command the pod status was `CrashLoopBackOff` (retrieved by `kubectl get pods`). Do you know what that is about?
- Is the plugin already supposed to work? This is the output I got:
```
lukp:~/kubectl-in-toto/example/verification $ kubectl in-toto pod/empty-test-6d64b7bf88-zldjs -k root_key.pub
[resolve] scanning pod: empty-test-6d64b7bf88-zldjs
[scan] resolved pod container as: docker://sha256:3ececd860353a9526c2015f700a8997fcaf2d30cc4eb9ccfa194f017dc0b7679 (empty:1.0.0).
	In-toto output follows:
<nil>
```


Btw. I fixed a few small things, while I went through the repo. Feel free to merge/cherry-pick...:
https://github.com/in-toto/kubectl-in-toto/compare/refactor-extend...lukpueh:refactor-extend-lp

SantiagoTorres commented:
I think the issue that you're running into is because either:

1. The container didn't build properly or
2. You didn't build with the docker socket inside the minikube cluster.

Can you check for either?

SantiagoTorres committed:
dep:in-toto: update to add paramsubs

SantiagoTorres committed:
in_toto: add image_id information

SantiagoTorres committed:
in_toto: cleanup, improve error reporting

SantiagoTorres committed:
example: improve example files

SantiagoTorres committed:
example/functionary/README: add --static to gcc

The a.out binary needs to be compiled statically for it to work on a
FROM scratch container

lukpueh commented on Makefile:
```diff
@@ -6,3 +6,6 @@ deploy:
 	@mkdir -vp $(KUBEPATH)
 	@go build -o $(KUBEPATH)/$(NAME)
 	echo plugin installed on $(KUBEPATH)/$(NAME)
+
```
Minor style nit: 
On my system this looks like...
```
lukp:~/in-toto/kubectl-in-toto $ make deploy
echo plugin installed on ~/.kube/plugins//'kubectl-in_toto'
plugin installed on /Users/lukp/.kube/plugins//kubectl-in_toto
```
Note the double path separators.

lukpueh commented on README.md:
```diff
@@ -1,25 +1,51 @@
+
+# Credit
 
 This was very heavily based off of [stefanprodan's kubectl-kubesec plugin](https://github.com/stefanprodan/kubectl-in-toto)
```
Ping https://github.com/in-toto/kubectl-in-toto/commit/c002e0af97661c8b72fb1a42341ef9435922bdcc

lukpueh commented on README.md:
```diff
@@ -1,25 +1,51 @@
+In order to scan a pod, you'd have to have the link metadata and the layout in
+your current folder. After passing the pod/podname argument, you can also use
+`-k` and `-l` in the same way as `in-toto-verify` to pass key and layout
+parameters.
```
It seems that the order of parameters is important, i.e. flags first and then positional:
```bash
kubectl in-toto [-k key -l layout] pod/[podname]
```

lukpueh commented on README.md:
```diff
@@ -1,25 +1,51 @@
+### Extensions
+
+The kubectl plugin uses parameter substitution to provide you with a
+`{IMAGE_ID}` parameter that you can substitute inside of your layouts. 
```
This is not yet used in the example,  right? You mentioned that you played around with it, do you have anything you'd like to share?

lukpueh commented on README.md:
```diff
@@ -1,25 +1,51 @@
+signed metadata files (you will need docker to build the container). If you're
+using minikube to run the example, I also suggest you expose the Docker socket
+before executing the functionary step so as to create the image inside the
+container.
```
Here's a single copy-pastable snippet to run through the example.
```bash
## Install kubectl plugin
# Also see https://kubernetes.io/docs/tasks/tools/install-kubectl/
make deploy
export PATH=~/.kube/plugins/:$PATH


# Start minikube
# Also see https://kubernetes.io/docs/tasks/tools/install-minikube/#install-minikube
minikube start


## Do the demo functionary work
cd example/functionary/

# NOTE: Compile on a system compatible with your `FROM scratch` container
gcc --static main.c

# Prepare for building docker image inside minikube
# Also see https://kubernetes.io/docs/setup/minikube/#use-local-images-by-re-using-the-docker-daemon
eval $(minikube docker-env)

# Build docker image generating signed in-toto link metadata
# Also see https://github.com/in-toto/in-toto/#installation
in-toto-run -k functionary -n build -p image_id -- docker build . -t empty:1.0.0 --iidfile image_id


## Spin up container image inside kubernetes cluster
# NOTE: Use `kubectl get pods` to see if the pods running
kubectl apply -f ../deployments/test.yml


## In-toto verification
cd ../verification/
# Cache pod name (required below)
pod_name=$(kubectl get pods --no-headers --output custom-columns=":metadata.name")

# Fetch in-toto link metadata from build functionary
# NOTE: If `root.layout` isn't there or has expired re-create using
# `../project-owner/create_layout.py` and copy it over too.
# `root_key.pub` is also required in this directory.
cp ../functionary/build.f5c8b97a.link .

# Verify that the software supply chain that resulted in the running pod was
# carried out according to its definition in a validly signed root.layout
kubectl in-toto -k root_key.pub -l root.layout pod/${pod_name}
```

Please let me know if I got it right.

lukpueh commented on pkg/in_toto/resources.go:
```diff
@@ -0,0 +1,56 @@
+
+    return result
+
+}
```
The resource type handling here looks a bit overly sophisticated, at least for now, where we only handle one resource type, and could call `ResolvePod` directly. 

What kind of resource types do we want to support in the future? And do they generalize enough so that a corresponding `Resolve*` function can always return a `[]ContainerSpec`?

lukpueh reviewed (commented):
Thanks for the updates, @SantiagoTorres! I guess the problem was indeed the missing compile flag, and probably also my building on an osx machine and trying to run it in a scratch container (silly me). I switched to building with the `--static` flag and on a debian box and got the demo running (see the sequence of commands I used below). :)

I also glanced over the code and I think it looks quite good (see a few comments below).  IMO it would be nice to run `go fmt` over it at some point and to also to add a bit of code documentation.

Let me know if you need any help.

lukpueh commented:
The scope of this PR has grown beyond switching to "native go library". I think it's okay to merge and address my comments and other things in subsequent PRs. 

@SantiagoTorres, would you mind going through my comments to take (mental) notes for things that might be worth revisiting? The example commands snippet, for instance, we might as well pipe into some `.md` file.

SantiagoTorres commented:
Ok! I'll make notes of these and make a new PR when these are addressed.
